File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public List<Integer> preorder(Node root) {
        List<Integer> ans = new ArrayList<>();
        if(root == null)
            return ans;
        Stack<Node> q = new Stack<>();
        q.push(root);
        
        while(!q.isEmpty()) {
            Node front = q.pop();
            ans.add(front.val);
            int n = front.children.size();
            for(int i=n-1;i>=0;i--) {
                q.push(front.children.get(i));
            }
        }
        return ans;
    }
}

File: 0589-n-ary-tree-preorder-traversal/0589-n-ary-tree-preorder-traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> ans;
        if(root == NULL)
            return ans;
        stack<Node*> q;
        q.push(root);
        
        while(!q.empty()) {
            Node* front = q.top();
            q.pop();
            ans.push_back(front->val);
            for(auto it = front -> children.rbegin();it!=front->children.rend();it++) {
                q.push(*it);
            }
        }
        return ans;
    }
};